// @terrestris/react-geo/dist/Util/TestUtil.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.TestUtil = void 0;

var React = _interopRequireWildcard(require("react"));

var _enzyme = require("enzyme");

var _View = _interopRequireDefault(require("ol/View"));

var _Map = _interopRequireDefault(require("ol/Map"));

var _Vector = _interopRequireDefault(require("ol/source/Vector"));

var _Vector2 = _interopRequireDefault(require("ol/layer/Vector"));

var _Feature = _interopRequireDefault(require("ol/Feature"));

var _Point = _interopRequireDefault(require("ol/geom/Point"));

var _MapBrowserEvent = _interopRequireDefault(require("ol/MapBrowserEvent"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * A set of some useful static helper methods.
 *
 * @class
 */
var TestUtil = /*#__PURE__*/_createClass(function TestUtil() {
  _classCallCheck(this, TestUtil);
});

exports.TestUtil = TestUtil;

_defineProperty(TestUtil, "mapDivId", 'map');

_defineProperty(TestUtil, "mapDivHeight", 256);

_defineProperty(TestUtil, "mapDivWidth", 256);

_defineProperty(TestUtil, "mountComponent", function (Component, props, options) {
  return (0, _enzyme.mount)( /*#__PURE__*/React.createElement(Component, props), options);
});

_defineProperty(TestUtil, "shallowComponent", function (Component, props, options) {
  return (0, _enzyme.shallow)( /*#__PURE__*/React.createElement(Component, props), options);
});

_defineProperty(TestUtil, "mountMapDiv", function () {
  var div = document.createElement('div');
  var style = div.style;
  style.position = 'absolute';
  style.left = '-1000px';
  style.top = '-1000px';
  style.width = TestUtil.mapDivWidth + 'px';
  style.height = TestUtil.mapDivHeight + 'px';
  div.id = TestUtil.mapDivId;
  document.body.appendChild(div);
  return div;
});

_defineProperty(TestUtil, "unmountMapDiv", function () {
  var div = document.querySelector("div#".concat(TestUtil.mapDivId));

  if (!div) {
    return;
  }

  var parent = div.parentNode;

  if (parent) {
    parent.removeChild(div);
  }

  div = null;
});

_defineProperty(TestUtil, "createMap", function (mapOpts) {
  var source = new _Vector["default"]();
  var layer = new _Vector2["default"]({
    source: source
  });
  var targetDiv = TestUtil.mountMapDiv();
  var defaultMapOpts = {
    target: targetDiv,
    layers: [layer],
    view: new _View["default"]({
      center: [829729, 6708850],
      resolution: 1,
      resolutions: mapOpts ? mapOpts.resolutions : undefined
    })
  };
  Object.assign(defaultMapOpts, mapOpts);
  var map = new _Map["default"](defaultMapOpts);
  map.renderSync();
  return map;
});

_defineProperty(TestUtil, "removeMap", function (map) {
  if (map instanceof _Map["default"]) {
    map.dispose();
  }

  TestUtil.unmountMapDiv();
});

_defineProperty(TestUtil, "simulatePointerEvent", function (_ref) {
  var map = _ref.map,
      type = _ref.type,
      x = _ref.x,
      y = _ref.y,
      optShiftKey = _ref.optShiftKey,
      dragging = _ref.dragging;
  var viewport = map.getViewport(); // Calculated in case body has top < 0 (test runner with small window).

  var position = viewport.getBoundingClientRect();
  var shiftKey = optShiftKey !== undefined ? optShiftKey : false;
  var event = new PointerEvent(type, {
    clientX: position.left + x + TestUtil.mapDivWidth / 2,
    clientY: position.top + y + TestUtil.mapDivHeight / 2,
    shiftKey: shiftKey
  });
  var olEvt = (0, _MapBrowserEvent["default"])(type, map, event, dragging);
  map.handleMapBrowserEvent(olEvt);
});

_defineProperty(TestUtil, "createVectorLayer", function (properties) {
  var source = new _Vector["default"]();
  var layer = new _Vector2["default"]({
    source: source
  });
  layer.setProperties(properties);
  return layer;
});

_defineProperty(TestUtil, "generatePointFeature", function () {
  var props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {
    ATTR_1: Math.random() * 100,
    ATTR_2: 'Borsigplatz 9',
    ATTR_3: 'Dortmund'
  };
  var coords = [Math.floor(Math.random() * 180) - 180, Math.floor(Math.random() * 90) - 90];
  var geom = new _Point["default"](coords);
  var feat = new _Feature["default"]({
    geometry: geom
  });
  feat.setProperties(props);
  return feat;
});

var _default = TestUtil;
exports["default"] = _default;
